tests/Unit/SafeMarkdownHelperTest.php

```diff
@@ -0,0 +1,108 @@
+        'hello earth',
+    ],
+    'escaped replacement' => [
+        ['hello :world', ['world' => '<strong>earth</strong>']],
+        'hello &lt;strong&gt;earth&lt;/strong&gt;',
+    ],
+    'escaped and unescaped replacement' => [
+        ['hello :world; good morning :!world', ['world' => '<strong>earth</strong>']],
+        'hello &lt;strong&gt;earth&lt;/strong&gt;; good morning <strong>earth</strong>',
+    ],
+    'replace URL' => [
+        ['hello <:url>', ['url' => 'http://example.com?a=1&b=2']],
+        'hello <a href="http://example.com?a=1&amp;b=2">http://example.com?a=1&amp;b=2</a>',
+    ],
+    'replace email' => [
+        ['hello <:email>', ['email' => 'test_1&2@example.com']],
+        'hello <a href="mailto:test_1&amp;2@example.com">test_1&amp;2@example.com</a>',
+    ],
+    'replace <:placeholder> without a URL or email' => [
+        ['hello <:world>', ['world' => '<strong>earth</strong>']],
+        'hello &lt;&lt;strong&gt;earth&lt;/strong&gt;&gt;',
+    ],
+    'replace entity encoded URL placeholder' => [
+        ['hello &lt;:url&gt;', ['url' => 'http://example.com?a=1&b=2']],
+        'hello <a href="http://example.com?a=1&amp;b=2">http://example.com?a=1&amp;b=2</a>',
+    ],
+    'string has html in it' => [
+        ['hello <strong>:world</strong>', ['world' => 'earth']],
+        'hello <strong>earth</strong>',
+    ],
+    'string has markdown in it' => [
+        ['hello **:world**', ['world' => 'earth']],
+        'hello **earth**',
+    ],
+]);
+
+test('safe markdown conversion', function ($args, $expected) {
+    $inlineMarkdownOutput = call_user_func_array('safe_inlineMarkdown', $args);
+    expect($inlineMarkdownOutput)->toBeInstanceOf(HtmlString::class);
+    expect(trim($inlineMarkdownOutput))->toEqual($expected);
+
+    $markdownOutput = call_user_func_array('safe_markdown', $args);
+    expect($inlineMarkdownOutput)->toBeInstanceOf(HtmlString::class);
+    expect(trim($markdownOutput))->toEqual("<p>{$expected}</p>");
+})->with([
+    'accepted markdown syntax' => [
+        ['_**NOTE:** Testing_'],
+        '<em><strong>NOTE:</strong> Testing</em>',
+    ],
+    'escape HTML' => [
+        ['Test <strong>HTML</strong> inline'],
+        'Test &lt;strong&gt;HTML&lt;/strong&gt; inline',
+    ],
+    'remove unsafe links' => [
+        ['Test [unsafe](javascript:alert("hello")) link'],
+        'Test <a>unsafe</a> link',
+    ],
+    'markdown unprocessed in replacement' => [
+        ['Injected :markdown', ['markdown' => '**[test](http://example.com)**']],
+        'Injected **[test](http://example.com)**',
+    ],
+    'escaped html in replacement' => [
+        ['Injected :markdown', ['markdown' => '<strong>should escape</strong>']],
+        'Injected &lt;strong&gt;should escape&lt;/strong&gt;',
+    ],
+    'Unescaped replacement' => [
+        ['Test <strong>[:text](:!url)</strong>', ['text' => '**my** <em>link</em>', 'url' => 'http://example.com?fruit=apple&condiment=cinamon']],
```

Comment:
I'm thinking maybe a different example here instead of a URL for the unescaped replacement. As in general, we want to use the escaping replacement for URLs.